actions/get-latest-workflow-artifact/src/main.ts

```diff
@@ -32,50 +33,96 @@ async function main() {
   });
 
   const headSha = data.head.sha;
-  // Now let's get all workflows associated with this sha.
-  const { data: runs } = await client.rest.actions.listWorkflowRuns({
-    repo: repoName,
-    owner: repoOwner,
-    head_sha: headSha,
-    workflow_id: workflowId,
-    status: "completed",
-  });
 
-  if (runs.workflow_runs.length <= 0) {
-    throw new Error(`No workflow runs found for sha ${headSha}`);
-  }
+  // Now let's get all workflows associated with this sha:
+  // We start with in-progress ones if those are to be considered.
+  let found = null;
 
-  const latestRun = runs.workflow_runs[0];
-  core.info(`Latest run: ${latestRun.id} <${latestRun.html_url}>`);
+  if (considerInProgress) {
+    found = await getLatestArtifact(
+      client,
+      repoOwner,
+      repoName,
+      workflowId,
+      headSha,
+      "in_progress",
+      artifactName,
+    );
+    core.setOutput("workflow-run-status", "in_progress");
+  }
 
-  // Now that we have the run we can get a list of all artifacts there:
-  const { data: artifacts } =
-    await client.rest.actions.listWorkflowRunArtifacts({
-      owner: repoOwner,
-      repo: repoName,
-      run_id: latestRun.id,
-    });
+  if (!found) {
+    found = await getLatestArtifact(
+      client,
+      repoOwner,
+      repoName,
+      workflowId,
+      headSha,
+      "completed",
+      artifactName,
+    );
+    core.setOutput("workflow-run-status", "completed");
+  }
 
-  const artifact = new DefaultArtifactClient();
-  const candidates = artifacts.artifacts.filter(
-    (artifact) => artifact.name == artifactName,
-  );
-  if (candidates.length <= 0) {
+  if (!found) {
     throw new Error(`No artifacts found with name ${artifactName}`);
   }
-
-  const response = await artifact.downloadArtifact(candidates[0].id, {
+  const { artifact: foundArtifact, run } = found;
+  const artifact = new DefaultArtifactClient();
+  const response = await artifact.downloadArtifact(foundArtifact.id, {
     path: path,
     findBy: {
       repositoryName: repoName,
       repositoryOwner: repoOwner,
-      workflowRunId: latestRun.id,
+      workflowRunId: run.id,
       token: ghToken,
     },
   });
   core.setOutput("artifact-download-path", response.downloadPath);
-  core.setOutput("artifact-id", candidates[0].id);
-  core.setOutput("workflow-run-id", latestRun.id);
+  core.setOutput("artifact-id", foundArtifact.id);
+  core.setOutput("workflow-run-id", run.id);
 }
 
+async function getLatestArtifact(
+  client,
+  repoOwner,
+  repoName,
+  workflowId,
+  headSha: string,
+  status: string,
+  artifactName: string,
+): Promise<{ artifact: any; run: any } | null> {
```

Comment:
These should be given better types!

If you add this at the top

```typescript
import type { components } from "@octokit/openapi-types";

type GetArtifactResponse = components["schemas"]["artifact"];
type GHClient = ReturnType<typeof github.getOctokit>;
type WorkflowJobStatus = components["schemas"]["job"]["status"];
type WorkflowRun = components["schemas"]["workflow-run"];
```

this could become

```suggestion
async function getLatestArtifact(
  client: GHClient,
  repoOwner: string,
  repoName: string,
  workflowId: string,
  headSha: string,
  status: WorkflowJobStatus,
  artifactName: string,
): Promise<{ artifact: GetArtifactResponse; run: WorkflowRun } | null> {
```